packages/upload-client/src/sharding.js

```diff
@@ -22,21 +23,30 @@ export class ShardingStream extends TransformStream {
     let shard = []
     /** @type {import('@ipld/unixfs').Block[] | null} */
     let readyShard = null
-    let size = 0
+    let shardBlockLength = 0
 
     super({
       async transform(block, controller) {
         if (readyShard != null) {
           controller.enqueue(await encode(readyShard))
           readyShard = null
         }
-        if (shard.length && size + block.bytes.length > shardSize) {
+
+        const blockLength = blockEncodingLength(block)
+        if (headerEncodingLength() + blockLength > shardSize) {
```

Comment:
Nit: I'd personally defined `const maxBlockLength = shardSize - headerEncodingLength()` on line 22 and then would have `if (blockLength > maxBlockLength)` here which would make logic here a lot more clear and avoid calculating on each block.